pkg/services/hivelocity/device/device.go

```diff
@@ -145,79 +124,51 @@ func (s *Service) Reconcile(ctx context.Context) (_ *ctrl.Result, err error) {
-
-	condition := conditions.Get(s.scope.HivelocityMachine, infrav1.DeviceReadyCondition)
-	if condition != nil &&
-		condition.Status == corev1.ConditionFalse &&
-		condition.Reason == infrav1.DeviceOffReason {
-		if time.Now().Before(condition.LastTransitionTime.Time.Add(deviceOffTimeout)) {
-			// Not yet timed out, try again to power on
-			if err := s.scope.HVClient.PowerOnDevice(ctx, device.DeviceId); err != nil {
-				return nil, fmt.Errorf("failed to power on device: %w", err)
-			}
-		} else {
-			// Timed out. Set failure reason
-			s.scope.SetError("reached timeout of waiting for machines that are switched off", capierrors.CreateMachineError)
-			return nil, nil
+func (s *Service) updateDevice(ctx context.Context, log logr.Logger, device *hv.BareMetalDevice) error {
+	log.V(1).Info("Updating machine")
+
+	// if the machine is already provisioned, return
+	if s.scope.Machine.Spec.ProviderID != nil {
+		// ensure ready state is set.
+		// This is required after move, because status is not moved to the target cluster.
+		s.scope.HivelocityMachine.Status.Ready = true
+		deviceID, err := hvutils.ProviderIDToDeviceID(*s.scope.Machine.Spec.ProviderID)
+		if err != nil {
+			return fmt.Errorf("[updateDevice] ProviderIDToDeviceID failed: %w", err)
 		}
-	} else { // todo: too complicated. Is there a way to avoid the "else"?
-		// No condition set yet. Try to power device on.
-		if err := s.scope.HVClient.PowerOnDevice(ctx, device.DeviceId); err != nil {
-			if hvclient.IsRateLimitExceededError(err) {
-				conditions.MarkTrue(s.scope.HivelocityMachine, infrav1.RateLimitExceeded)
-				record.Event(s.scope.HivelocityMachine,
-					"RateLimitExceeded",
-					"exceeded rate limit with calling hivelocity function PowerOnDevice",
-				)
-			}
-			return nil, fmt.Errorf("failed to power on device: %w", err)
+		exists, err := hvutils.DeviceExists(ctx, s.scope.HVClient, deviceID)
+		if err != nil {
+			return fmt.Errorf("[updateDevice] DeviceExists failed: %w", err)
 		}
-		conditions.MarkFalse(
-			s.scope.HivelocityMachine,
-			infrav1.DeviceReadyCondition,
-			infrav1.DeviceOffReason,
-			clusterv1.ConditionSeverityInfo,
-			"device is switched off",
-		)
+		if exists {
+			conditions.MarkTrue(s.scope.HivelocityMachine, infrav1.DeviceReadyCondition)
+			// Setting address is required after move, because Status field is not retained during move.
+			setMachineAddress(s.scope.HivelocityMachine, device)
+		} else {
+			conditions.MarkFalse(s.scope.HivelocityMachine,
+				infrav1.DeviceReadyCondition,
+				infrav1.DeviceDeletedReason,
+				clusterv1.ConditionSeverityError,
+				fmt.Sprintf("device %d does not exists anymore", device.DeviceId))
+		}
+		return nil
 	}
 
-	// Try again in 30 sec.
-	return &reconcile.Result{RequeueAfter: 30 * time.Second}, nil
-}
-
-func (s *Service) createDevice(ctx context.Context) (*hv.BareMetalDevice, error) {
-	// get userData
+	// question: this is from server.go. In CAPD conditions get updated.
```

Comment:
Does CAPD need userdata as they don't need to provision servers?

Reply:
capd uses userdata and cloud-init, too. Here some code:
```
	switch format {
	case bootstrapv1.CloudConfig:
		commands, err = cloudinit.RawCloudInitToProvisioningCommands(cloudConfig)
	case bootstrapv1.Ignition:
		commands, err = ignition.RawIgnitionToProvisioningCommands(cloudConfig)
	default:
		return fmt.Errorf("unknown provisioning format %q", format)
	}
```
file: cluster-api/test/infrastructure/docker/internal/docker/machine.go

Reply:
Does this answer your question?